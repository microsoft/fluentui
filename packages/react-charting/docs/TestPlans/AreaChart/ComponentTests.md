**Area Chart – Component test plan**

**Sub-components: Area, x-axis, y-axis and Legend**

1. **Line: Area data, Area color (multi colors), Area label (show/hide)**
1. **Legends: show/hide legends, highlight the corresponding Area on legend hover**
1. **Callout: Default/custom callout**
1. **Labels: x-Axis labels, y-Axis labels**

|                            **Test steps**                             |                                                 **Validation**                                                  | **Tool used** |
| :-------------------------------------------------------------------: | :-------------------------------------------------------------------------------------------------------------: | :-----------: |
|                      Test 1: [Snapshot testing]                       |                                                                                                                 |               |
|           - With only data prop, numerical data on x-axis.            |                                          Renders Area chart correctly                                           |      RTL      |
|              - With only data prop, date data on x-axis.              |                                          Renders Area chart correctly                                           |      RTL      |
|             - With only data prop, string data on x-axis.             |                                          Should not render area chart                                           |    Pending    |
|                 - With HideLegend prop set to “true”                  |                                               Should hide legends                                               |    Enzyme     |
|                 - With HideTooltip prop set to “true”                 |                                        Should hide the tooltip in chart                                         |    Enzyme     |
|             - With EnabledLegendsWrapLines set to “true”              |    Should enable the legends to wrap lines if there is not enough space to show all legends on a single line    |    Enzyme     |
|              - With ShowXAxisLablesTooltip set to “true”              |                         Should truncate x axis labels and show tooltip on x axis labels                         |    Enzyme     |
|                 - With WrapXAxisLables set to “true”                  |                                         Should wrap x axis label values                                         |    Enzyme     |
|                  - With yAxisTickFormat set to “%d”                   |                      <p>Should render the y-axis ticks in the format specified</p><p></p>                       |    Enzyme     |
|                          - With single point                          |                                    Should render Area chat with single point                                    |    Enzyme     |
|                         - With Default color                          |                         Should render with default colors when line color not provided                          |    Enzyme     |
|                        - With specific colors                         |                                    Should render areas with specified colors                                    |      RTL      |
|                - With optimizeLargeData set to “true”                 |                            Should not render circles when optimizeLargeData is true                             |    Enzyme     |
|                      Test 2: Basic props testing                      |                                                                                                                 |               |
|                    - HideLegend prop set to “true”                    |                                 Should not mount legend when hideLegend is true                                 |    Enzyme     |
|                   - HideLegend prop set to “false”                    |                                  Should mount legend when hideLegend is false                                   |    Enzyme     |
|                   - HideTooltip prop set to “true”                    |                                Should not mount callout when hideTootip is true                                 |    Enzyme     |
|                   - HideTooltip prop set to “false”                   |                                  Should mount callout when hideTootip is false                                  |    Enzyme     |
|              - onRenderCalloutPerStack prop is not given              |                                    Should not render onRenderCalloutPerStack                                    |    Enzyme     |
|                - onRenderCalloutPerDataPoint is given                 |                                    Should render onRenderCalloutPerDataPoint                                    |    Enzyme     |
|              - onRenderCalloutPerDataPoint is not given               |                                  Should not render onRenderCalloutPerDataPoint                                  |    Enzyme     |
|            Test 3: Render calling with respective to props            |                                                                                                                 |               |
| - No prop changes: Mount Area chart and then set the same props again |                                  Render function should have been called twice                                  |    Enzyme     |
|     - Prop changes: Mount Area chart and then set some other prop     |                                  Render function should have been called twice                                  |    Enzyme     |
|                         Test 4: Mouse events                          |                                                                                                                 |               |
|                         - Mouse over on Area                          |                                  Should render callout correctly on mouseover                                   |    Enzyme     |
|               - Mouse move from one area to other area                |                                  Should render callout correctly on mouse move                                  |    Enzyme     |
|            - Mouse over on a area with customized callout             |                                  Should render customized callout on mouseover                                  |    Enzyme     |
|       - Customized callout on area from one area to other area        |                             Should render customized callout for stack on mouseover                             |    Enzyme     |
|     Test 5: Render empty chart aria label div when chart is empty     |                                                                                                                 |               |
|               - Area chart mounted with non-empty data                |                                     No empty chart aria label div rendered                                      |    Enzyme     |
|                 - Area chart mounted with empty data                  |                                       Empty chart aria label div rendered                                       |    Enzyme     |
|                   Test 6: [Sub-Component]: Legends                    |                                                                                                                 |               |
|                    - Hover mouse over area legends                    |                                  Should reduce the opacity of the other Areas                                   |      RTL      |
|                    - Hover mouse over area legends                    | Should reduce the opacity of the other lines in area chart and opacity of the selected area line should be zero |      RTL      |
|                    - Hover mouse over area legends                    |                                 Should reduce the opacity of the other legends                                  |      RTL      |
|                    - Single mouse click on legends                    |                         Should select legend on single mouse click on respective legend                         |      RTL      |
|                    - Double mouse click on legends                    |                        Should deselect legend on double mouse click on respective legend                        |      RTL      |
|                   Test 7: [Sub-Component]: Callout                    |                                                                                                                 |               |
|                   - Hover mouse over a single area                    |                                 Should show the default callout over that Area                                  |      RTL      |
|                   - Hover mouse over a stacked area                   |                             Should show the default stacked callout over that Area                              |      RTL      |
|         - Specify custom callout and hover mouse over a Area          |                                  Should show the custom callout over that Area                                  |      RTL      |
|                Test 8: [Sub-Component]: x-axis labels                 |                                                                                                                 |               |
|                       - Truncate x-axis labels                        |                               Should show the x-axis labels tooltip when hovered                                |      RTL      |
|                        - Rotate x-axis labels                         |                                  Should rotate the x-axis labels by 39 degrees                                  |      RTL      |
|              Test 9: [Sub-Component]: Screen resolution               |                                                                                                                 |               |
|              - Increase the screen resolution (zoom in)               |                                       Should remain unchanged on zoom in                                        |      RTL      |
|              - Decrease the screen resolution (zoom out)              |                                       Should remain unchanged on zoom out                                       |      RTL      |
|                 Test 10: Theme changed to Dark Theme                  |                                           Should reflect theme change                                           |      RTL      |
