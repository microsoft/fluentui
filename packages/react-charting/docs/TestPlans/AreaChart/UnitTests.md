###**Unit Test Plan â€“ Area Chart**

|                                                       **Test steps**                                                       |                                                  **Validation**                                                  | **Tool used** |
| :------------------------------------------------------------------------------------------------------------------------: | :--------------------------------------------------------------------------------------------------------------: | :-----------: |
|                                         **1. Unit test for \_getOpacity function**                                         |                                                                                                                  |               |
|                                             i. Chart data is not multi-stacked                                             |                       Should return fill-opacity of 0.7 if chart data is not multi-stacked                       |      RTL      |
|                                              ii. Chart data is multi-stacked                                               |                           Should return fill-opacity of 0.7 if chart is multi-stacked                            |      RTL      |
|                                    iii. Legend is highlighted and data is multi-stacked                                    |               Should return fill-opacity of 0.7 if legend is highlighted and data is multi-stacked               |      RTL      |
|                         iv. Legend 1 is highlighted but legend 2 is not and data is multi-stacked                          |    Should return fill-opacity of 0.1 if legend 1 is highlighted but legend 2 is not and data is multi-stacked    |      RTL      |
|                                       **2. Unit test for \_getLineOpacity function**                                       |                                                                                                                  |               |
|                                             i. Chart data is not multi-stacked                                             |                                  Should return 1 if chart is not multi-stacked                                   |      RTL      |
|                                ii. Chart data is multi-stacked and no legend is highlighted                                |                     Should return 0.3 if chart is multi-stacked and no legend is highlighted                     |      RTL      |
|                                   iii. Callout is visible and data is not multi-stacked                                    |                       Should return 1 if callout is visible and data is not multi-stacked                        |      RTL      |
|                     iv. Legend 1 is highlighted, legend 2 is not highlighted and data is multi-stacked                     |  Should return opacity of 0.1 if legend 1 is highlighted, legend 2 is not highlighted and data is multi-stacked  |      RTL      |
|                          v. Legend 1 is highlighted, legend 2 is active and data is multi-stacked                          |       Should return opacity of 0 if legend 1 is highlighted, legend 2 is active and data is multi-stacked        |      RTL      |
|                                     **3. Unit test \_updateCircleFillColor function**                                      |                                                                                                                  |               |
|                             i. Neither the nearest circle nor the active point is highlighted                              |          Should return the line color if neither the nearest circle nor the active point is highlighted          |      RTL      |
|                            ii. The nearest circle is highlighted and the circle is not clicked                             |              Should return white if the nearest circle is highlighted and the circle is not clicked              |      RTL      |
|                              iii. The nearest circle is highlighted and the circle is clicked                              |           Should return the line color if the nearest circle is highlighted and the circle is clicked            |      RTL      |
|                             iv. The active point is highlighted and the circle is not clicked                              |               Should return white if the active point is highlighted and the circle is not clicked               |      RTL      |
|                                v. The active point is highlighted but the circle is clicked                                |            Should return the line color if the active point is highlighted but the circle is clicked             |      RTL      |
|                                      **4. Unit test for \_getCircleRadius function**                                       |                                                                                                                  |               |
|                         i. isCircleClicked is true and nearestCircleToHighlight matches xDataPoint                         |            Should return 1 if isCircleClicked is true and nearestCircleToHighlight matches xDataPoint            |      RTL      |
|                                             ii. ActivePoint matches the circle                                             |                           Should return circleRadius if activePoint matches the circle                           |      RTL      |
| iii. isCircleClicked is false, active point and circle do not match and nearestCircleToHighlight does not match xDataPoint |    Should return 0 if !isCircleClicked, (activePoint != circle) and (nearestCircleToHighlight != xDataPoint)     |      RTL      |
|                     iv. isCircleClicked is true but nearestCircleToHighlight does not match xDataPoint                     |        Should return 0 if isCircleClicked is true but nearestCircleToHighlight does not match xDataPoint         |      RTL      |
|                      v. nearestCircleToHighlight does not match xDataPoint and activePoint is not set                      |         Should return 0 if nearestCircleToHighlight does not match xDataPoint and activePoint is not set         |      RTL      |
|                         vi. nearestCircleToHighlight matches xDataPoint and activePoint is not set                         |       Should return circleRadius if nearestCircleToHighlight matches xDataPoint and activePoint is not set       |      RTL      |
|                    vii. nearestCircleToHighlight matches xDataPoint, activePoint matches another circle                    |  Should return circleRadius if nearestCircleToHighlight matches xDataPoint, activePoint matches another circle   |      RTL      |
|                                      **5. Unit test for \_addDefaultColors function**                                      |                                                                                                                  |               |
|                                           i. Passing data without default color                                            |                          Should return an array with the same length as the input array                          |      RTL      |
|                                           ii. Passing data without default color                                           |                            Should add default colors to each item in the input array                             |      RTL      |
|                                            iii. Passing data with default color                                            |                                    Should use the provided color if it exists                                    |      RTL      |
|                                iv. isInverted property is true and colors are not provided                                 |           Should use the inverted theme if the isInverted property is true and colors are not provided           |      RTL      |
|                                   v. isInverted property is true and colors are provided                                   |             Should use the inverted theme if the isInverted property is true and colors are provided             |      RTL      |
|                                          **6. Unit test \_getAriaLabel function**                                          |                                                                                                                  |               |
|                                i. Point contains both xAxisCalloutData and yAxisCalloutData                                |           Should return the correct aria label for a point with xAxisCalloutData and yAxisCalloutData            |      RTL      |
|                                        ii. Point with formatted x value and y value                                        |               Should return the correct aria label for a point with formatted x value and y value                |      RTL      |
|                                   iii. Point with formatted x value and yAxisCalloutData                                   |           Should return the correct aria label for a point with formatted x value and yAxisCalloutData           |      RTL      |
|                                   iv. Point with formatted x value and xAxisCalloutData                                    |           Should return the correct aria label for a point with formatted x value and xAxisCalloutData           |      RTL      |
|                          v. Point with callOutAccessibilityData but no xAxis or yAxis CalloutData                          | Should return the correct aria label for a point with callOutAccessibilityData but no xAxis or yAxis CalloutData |      RTL      |
