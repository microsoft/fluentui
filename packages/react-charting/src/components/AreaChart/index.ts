export * from './AreaChart';
export * from './AreaChart.types';
