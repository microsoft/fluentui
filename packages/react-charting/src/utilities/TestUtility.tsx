/* eslint-disable @typescript-eslint/no-explicit-any */
import { act, queryAllByAttribute, render, waitFor } from '@testing-library/react';
import * as React from 'react';

export const getById = queryAllByAttribute.bind(null, 'id');
export const getByClass = queryAllByAttribute.bind(null, 'class');

export const testWithoutWait = (
  description: string,
  component: any,
  props: any,
  testFunction: (container: HTMLElement) => void,
  testFunctionBeforeWait?: () => void,
) => {
  test(description, () => {
    const { container } = render(React.createElement(component, (props = { ...props })));
    testFunctionBeforeWait !== undefined && testFunctionBeforeWait();
    testFunction(container);
  });
};

export const testWithWait = (
  description: string,
  component: any,
  props: any,
  testFunction: (container: HTMLElement) => void,
) => {
  test(description, async () => {
    const { container } = render(React.createElement(component, (props = { ...props })));
    await waitFor(() => {
      testFunction(container);
    });
  });
};

export const testScreenResolution = (testFunction: () => void) => {
  const originalInnerWidth = global.innerWidth;
  const originalInnerHeight = global.innerHeight;
  afterEach(() => {
    global.innerWidth = originalInnerWidth;
    global.innerHeight = originalInnerHeight;
    act(() => {
      global.dispatchEvent(new Event('resize'));
    });
  });
  test('Should remain unchanged on zoom in', () => {
    testFunction();
    global.innerWidth = window.innerWidth / 2;
    global.innerHeight = window.innerHeight / 2;
    act(() => {
      global.dispatchEvent(new Event('resize'));
    });
  });
  test('Should remain unchanged on zoom out', () => {
    testFunction();
    global.innerWidth = window.innerWidth * 2;
    global.innerHeight = window.innerHeight * 2;
    act(() => {
      global.dispatchEvent(new Event('resize'));
    });
  });
};
