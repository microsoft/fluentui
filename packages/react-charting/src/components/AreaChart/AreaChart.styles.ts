import { IAreaChartStyleProps, IAreaChartStyles } from './AreaChart.types';

export const getStyles = (props: IAreaChartStyleProps): IAreaChartStyles => {
  return {
    tooltip: {
      ...props.theme!.fonts.medium,
      display: 'flex',
      flexDirection: 'column',
      padding: '8px',
      position: 'absolute',
      textAlign: 'center',
      top: '0px',
      background: props.theme!.semanticColors.bodyBackground,
      borderRadius: '2px',
      pointerEvents: 'none',
    },
  };
};
