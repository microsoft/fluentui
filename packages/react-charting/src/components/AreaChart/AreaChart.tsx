import * as React from 'react';
import { styled } from '@fluentui/react/lib/Utilities';
import { IAreaChartProps, IAreaChartStyleProps, IAreaChartStyles } from '../../index';
import { AreaChartBase } from '../AreaChart/AreaChart.base';
import { getStyles } from './AreaChart.styles';

// Create a AreaChart variant which uses these default styles and this styled subcomponent.
export const AreaChart: React.FunctionComponent<IAreaChartProps> = styled<
  IAreaChartProps,
  IAreaChartStyleProps,
  IAreaChartStyles
>(AreaChartBase, getStyles);
