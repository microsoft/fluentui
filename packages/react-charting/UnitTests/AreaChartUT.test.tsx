import { screen, fireEvent } from '@testing-library/react';
import { AreaChart } from '../src/components/AreaChart/index';

import { getById, testWithoutWait } from '../src/utilities/TestUtility.test';
import { DarkTheme } from '@fluentui/theme-samples';
const env = require('../config/tests');

const runTest = env === 'TEST' ? describe : describe.skip;

const chart1Points = [
  {
    x: 20,
    y: 9,
  },
  {
    x: 25,
    y: 14,
  },
  {
    x: 30,
    y: 14,
  },
  {
    x: 35,
    y: 23,
  },
  {
    x: 40,
    y: 20,
  },
  {
    x: 45,
    y: 31,
  },
  {
    x: 50,
    y: 29,
  },
  {
    x: 55,
    y: 27,
  },
  {
    x: 60,
    y: 37,
  },
  {
    x: 65,
    y: 51,
  },
];

const chart2Points = [
  {
    x: 20,
    y: 21,
  },
  {
    x: 25,
    y: 25,
  },
  {
    x: 30,
    y: 10,
  },
  {
    x: 35,
    y: 10,
  },
  {
    x: 40,
    y: 14,
  },
  {
    x: 45,
    y: 18,
  },
  {
    x: 50,
    y: 9,
  },
  {
    x: 55,
    y: 23,
  },
  {
    x: 60,
    y: 7,
  },
  {
    x: 65,
    y: 55,
  },
];

const chart3Points = [
  {
    x: 20,
    y: 30,
  },
  {
    x: 25,
    y: 35,
  },
  {
    x: 30,
    y: 33,
  },
  {
    x: 35,
    y: 40,
  },
  {
    x: 40,
    y: 10,
  },
  {
    x: 45,
    y: 40,
  },
  {
    x: 50,
    y: 34,
  },
  {
    x: 55,
    y: 40,
  },
  {
    x: 60,
    y: 60,
  },
  {
    x: 65,
    y: 40,
  },
];

const chartPointsMultiStacked = [
  {
    legend: 'legend1',
    data: chart1Points,
    color: 'green',
  },
  {
    legend: 'legend2',
    data: chart2Points,
    color: 'yellow',
  },
  {
    legend: 'legend3',
    data: chart3Points,
    color: 'blue',
  },
];

const chartPoints = [
  {
    legend: 'legend1',
    data: chart1Points,
    color: 'green',
  },
];

const chartPointsDarkTheme = [
  {
    legend: 'legend1',
    data: chart1Points,
    color: DarkTheme.palette.themePrimary,
  },
];

const chartPointsWithoutDefaultColor = [
  {
    legend: 'legend1',
    data: chart1Points,
  },
];

export const chartDataMultiStacked = {
  chartTitle: 'Area chart multiple example',
  lineChartData: chartPointsMultiStacked,
};

export const chartData = {
  chartTitle: 'Area chart example',
  lineChartData: chartPoints,
};

export const chartDataWithoutDefaultColor = {
  chartTitle: 'Area chart without default color example',
  lineChartData: chartPointsWithoutDefaultColor,
};

export const chartDataDarkTheme = {
  chartTitle: 'Area chart dark theme example',
  lineChartData: chartPointsDarkTheme,
};

runTest('_getOpacity', () => {
  testWithoutWait(
    'Should return fill-opacity of 0.7 if chart is not multi-stacked',
    AreaChart,
    { data: chartData },
    container => {
      const areas = getById(container, /graph-areaChart/i);
      expect(areas.length).toBe(1);
      expect(areas[0].getAttribute('fill-opacity')).toBe('0.7');
    },
  );

  testWithoutWait(
    'Should return fill-opacity of 0.7 if chart is multi-stacked',
    AreaChart,
    { data: chartDataMultiStacked },
    container => {
      const areas = getById(container, /graph-areaChart/i);
      expect(areas.length).toBe(3);
      expect(areas[0].getAttribute('fill-opacity')).toBe('0.7');
    },
  );

  testWithoutWait(
    'Should return fill-opacity of 0.7 if legend is highlighted and data is multi-stacked',
    AreaChart,
    { data: chartDataMultiStacked },
    container => {
      const legend = screen.queryByText('legend1');
      expect(legend).toBeDefined();
      fireEvent.mouseOver(legend!);
      // Assert
      const areas = getById(container, /graph-areaChart/i);
      expect(areas[0].getAttribute('fill-opacity')).toEqual('0.7');
    },
  );

  testWithoutWait(
    'Should return fill-opacity of 0.1 if legend 1 is highlighted but legend 2 is not and data is multi-stacked',
    AreaChart,
    { data: chartDataMultiStacked },
    container => {
      const legend1 = screen.queryByText('legend1');
      expect(legend1).toBeDefined();
      fireEvent.mouseOver(legend1!);
      // Assert
      const areas = getById(container, /graph-areaChart/i);
      expect(areas[1].getAttribute('fill-opacity')).toEqual('0.1');
    },
  );
});

runTest('_getLineOpacity', () => {
  testWithoutWait('Should return 1 if chart is not multi-stacked', AreaChart, { data: chartData }, container => {
    const areas = getById(container, /line-areaChart/i);
    expect(areas.length).toBe(1);
    expect(areas[0].getAttribute('opacity')).toBe('1');
  });

  testWithoutWait(
    'Should return 0.3 if chart is multi-stacked and no legend is highlighted',
    AreaChart,
    { data: chartDataMultiStacked },
    container => {
      const areas = getById(container, /line-areaChart/i);
      expect(areas.length).toBe(3);
      expect(areas[0].getAttribute('opacity')).toBe('0.3');
    },
  );

  testWithoutWait(
    'Should return 1 if callout is visible and data is not multi-stacked',
    AreaChart,
    { data: chartData },
    container => {
      // Arrange
      const areas = getById(container, /line-areaChart/i);
      expect(areas.length).toBe(1);
      fireEvent.mouseOver(areas[0]);
      // Assert
      expect(getById(container, /toolTipcallout/i)).toBeDefined();
      expect(areas[0].getAttribute('opacity')).toBe('1');
    },
  );

  testWithoutWait(
    'Should return opacity of 0.1 if legend 1 is highlighted, legend 2 is not highlighted and data is multi-stacked',
    AreaChart,
    { data: chartDataMultiStacked },
    container => {
      const legend1 = screen.queryByText('legend1');
      expect(legend1).toBeDefined();
      fireEvent.mouseOver(legend1!);
      // Assert
      const areas = getById(container, /line-areaChart/i);
      expect(areas.length).toBe(3);
      expect(areas[1].getAttribute('opacity')).toEqual('0.1');
    },
  );

  testWithoutWait(
    'Should return opacity of 0 if legend 1 is highlighted, legend 2 is active and data is multi-stacked',
    AreaChart,
    { data: chartDataMultiStacked },
    container => {
      const legend1 = screen.queryByText('legend1');
      expect(legend1).toBeDefined();
      fireEvent.mouseOver(legend1!);
      const legend2 = screen.queryByText('legend2');
      expect(legend2).toBeDefined();
      fireEvent.click(legend2!);
      // Assert
      const areas = getById(container, /line-areaChart/i);
      expect(areas.length).toBe(3);
      expect(areas[1].getAttribute('opacity')).toEqual('0');
    },
  );
});

runTest('_updateCircleFillColor', () => {
  testWithoutWait(
    'Should return the line color if neither the nearest circle nor the active point is highlighted',
    AreaChart,
    { data: chartData },
    container => {
      const points = getById(container, /circle/i);
      expect(points).toHaveLength(10);
      // Assert
      points.forEach(point => {
        expect(point.getAttribute('fill')).toEqual('green');
      });
    },
  );

  testWithoutWait(
    'Should return white if the nearest circle is highlighted and the circle is not clicked',
    AreaChart,
    { data: chartData },
    container => {
      const points = getById(container, /circle/i);
      expect(points).toHaveLength(10);
      fireEvent.mouseOver(points[6]);
      expect(points[6].getAttribute('fill')).toEqual('#ffffff');
    },
  );

  testWithoutWait(
    'Should return the line color if the nearest circle is highlighted and the circle is clicked',
    AreaChart,
    { data: chartData },
    container => {
      const points = getById(container, /circle/i);
      expect(points).toHaveLength(10);
      fireEvent.mouseOver(points[0]);
      fireEvent.click(points[0]);
      expect(points[0].getAttribute('fill')).toEqual('green');
    },
  );

  testWithoutWait(
    'Should return white if the active point is highlighted and the circle is not clicked',
    AreaChart,
    { data: chartData },
    container => {
      const points = getById(container, /circle/i);
      expect(points).toHaveLength(10);
      fireEvent.focus(points[0]);
      expect(points[0].getAttribute('fill')).toEqual('#ffffff');
    },
  );

  testWithoutWait(
    'Should return the line color if the active point is highlighted but the circle is clicked',
    AreaChart,
    { data: chartData },
    container => {
      const points = getById(container, /circle/i);
      expect(points).toHaveLength(10);
      fireEvent.mouseOver(points[0]);
      fireEvent.click(points[0]);
      fireEvent.mouseOver(points[1]);
      fireEvent.click(points[1]);
      expect(points[1].getAttribute('fill')).toEqual('green');
    },
  );
});

runTest('_getCircleRadius', () => {
  testWithoutWait(
    'Should return 1 if isCircleClicked is true and nearestCircleToHighlight matches xDataPoint',
    AreaChart,
    { data: chartData },
    container => {
      const points = getById(container, /circle/i);
      expect(points).toHaveLength(10);
      fireEvent.mouseOver(points[1]);
      fireEvent.click(points[0]);
      // nearest circle to highlight has index 6
      // activePoint does not matche the point circle whose radius is checked
      expect(points[6].getAttribute('r')).toEqual('1');
    },
  );

  testWithoutWait(
    'Should return circleRadius if activePoint matches the circle',
    AreaChart,
    { data: chartData },
    container => {
      const points = getById(container, /circle/i);
      expect(points).toHaveLength(10);
      fireEvent.focus(points[0]);
      // activePoint matches the point circle whose radius is checked
      expect(points[0].getAttribute('r')).toEqual('8');
    },
  );

  testWithoutWait(
    'Should return 0 if !isCircleClicked, (activePoint != circle) and (nearestCircleToHighlight != xDataPoint)',
    AreaChart,
    { data: chartData },
    container => {
      const points = getById(container, /circle/i);
      expect(points).toHaveLength(10);
      fireEvent.focus(points[0]);
      expect(points[1].getAttribute('r')).toEqual('0');
    },
  );

  testWithoutWait(
    'Should return 0 if isCircleClicked is true but nearestCircleToHighlight does not match xDataPoint',
    AreaChart,
    { data: chartData },
    container => {
      const points = getById(container, /circle/i);
      expect(points).toHaveLength(10);
      fireEvent.click(points[0]);
      fireEvent.mouseOver(points[1]);
      expect(points[2].getAttribute('r')).toEqual('0');
    },
  );

  testWithoutWait(
    'Should return 0 if nearestCircleToHighlight does not match xDataPoint and activePoint is not set',
    AreaChart,
    { data: chartData },
    container => {
      const points = getById(container, /circle/i);
      expect(points).toHaveLength(10);
      fireEvent.mouseOver(points[1]);
      expect(points[2].getAttribute('r')).toEqual('0');
    },
  );

  testWithoutWait(
    'Should return circleRadius if nearestCircleToHighlight matches xDataPoint and activePoint is not set',
    AreaChart,
    { data: chartData },
    container => {
      const points = getById(container, /circle/i);
      expect(points).toHaveLength(10);
      fireEvent.mouseOver(points[6]);
      expect(points[6].getAttribute('r')).toEqual('8');
    },
  );

  testWithoutWait(
    'Should return circleRadius if nearestCircleToHighlight matches xDataPoint, activePoint matches another circle',
    AreaChart,
    { data: chartData },
    container => {
      const points = getById(container, /circle/i);
      expect(points).toHaveLength(10);
      fireEvent.focus(points[0]);
      fireEvent.mouseOver(points[6]);
      expect(points[6].getAttribute('r')).toEqual('8');
    },
  );
});

runTest('_addDefaultColors', () => {
  testWithoutWait(
    'Should return an array with the same length as the input array',
    AreaChart,
    { data: chartDataWithoutDefaultColor },
    container => {
      const points = getById(container, /circle/i);
      expect(points).toHaveLength(10);
      points.forEach(point => {
        expect(point).toHaveAttribute('fill');
        expect(point.getAttribute('fill')).toEqual('#637cef');
      });
    },
  );

  testWithoutWait('Should use the provided color if it exists', AreaChart, { data: chartData }, container => {
    const points = getById(container, /circle/i);
    expect(points).toHaveLength(10);
    points.forEach(point => {
      expect(point).toHaveAttribute('fill');
      expect(point.getAttribute('fill')).toEqual('green');
    });
  });

  testWithoutWait(
    'Should use the inverted theme if the isInverted property is true and colors are not provided',
    AreaChart,
    { data: chartDataWithoutDefaultColor, theme: { ...DarkTheme, isInverted: true } },
    container => {
      const points = getById(container, /circle/i);
      expect(points).toHaveLength(10);
      points.forEach(point => {
        expect(point).toHaveAttribute('fill');
        expect(point.getAttribute('fill')).toEqual('#637cef');
      });
    },
  );

  testWithoutWait(
    'Should use the inverted theme if the isInverted property is true and colors are provided',
    AreaChart,
    { data: chartDataDarkTheme, theme: { ...DarkTheme, isInverted: true } },
    container => {
      const points = getById(container, /circle/i);
      expect(points).toHaveLength(10);
      points.forEach(point => {
        expect(point).toHaveAttribute('fill');
        expect(point.getAttribute('fill')).toEqual('#2899f5');
      });
    },
  );
});

runTest('_getAriaLabel', () => {
  testWithoutWait(
    'Should return the correct aria label for a point with xAxisCalloutData and yAxisCalloutData',
    AreaChart,
    {
      data: {
        chartTitle: 'Area chart arial label example',
        lineChartData: [
          {
            legend: 'Legend 1',
            data: [
              {
                x: new Date(2022, 0, 1),
                y: 10,
                xAxisCalloutData: 'Jan 1, 2022',
                yAxisCalloutData: '10 units',
                callOutAccessibilityData: {
                  ariaLabel: 'Custom aria label',
                },
              },
            ],
          },
        ],
      },
    },
    container => {
      const points = getById(container, /circle/i);
      expect(points).toHaveLength(1);
      expect(points[0].getAttribute('aria-label')).toEqual('Custom aria label');
    },
  );

  testWithoutWait(
    'Should return the correct aria label for a point with formatted x value and y value',
    AreaChart,
    {
      data: {
        chartTitle: 'Area chart arial label example',
        lineChartData: [
          {
            legend: 'Legend 1',
            data: [
              {
                x: new Date(2022, 0, 1),
                y: 10,
              },
            ],
          },
        ],
      },
    },
    container => {
      const points = getById(container, /circle/i);
      expect(points).toHaveLength(1);
      expect(points[0].getAttribute('aria-label')).toEqual('1/1/2022, 12:00:00 AM. Legend 1, 10.');
    },
  );

  testWithoutWait(
    'Should return the correct aria label for a point with formatted x value and yAxisCalloutData',
    AreaChart,
    {
      data: {
        chartTitle: 'Area chart arial label example',
        lineChartData: [
          {
            legend: 'Legend 1',
            data: [
              {
                x: new Date(2022, 0, 1),
                y: 10,
                yAxisCalloutData: '10 units',
              },
            ],
          },
        ],
      },
    },
    container => {
      const points = getById(container, /circle/i);
      expect(points).toHaveLength(1);
      expect(points[0].getAttribute('aria-label')).toEqual('1/1/2022, 12:00:00 AM. Legend 1, 10 units.');
    },
  );

  testWithoutWait(
    'Should return the correct aria label for a point with formatted x value and xAxisCalloutData',
    AreaChart,
    {
      data: {
        chartTitle: 'Area chart arial label example',
        lineChartData: [
          {
            legend: 'Legend 1',
            data: [
              {
                x: new Date(2022, 0, 1),
                y: 10,
                xAxisCalloutData: 'Jan 1, 2022',
              },
            ],
          },
        ],
      },
    },
    container => {
      const points = getById(container, /circle/i);
      expect(points).toHaveLength(1);
      expect(points[0].getAttribute('aria-label')).toEqual('Jan 1, 2022. Legend 1, 10.');
    },
  );

  testWithoutWait(
    'Should return the correct aria label for a point with callOutAccessibilityData but no xAxis or yAxis CalloutData',
    AreaChart,
    {
      data: {
        chartTitle: 'Area chart arial label example',
        lineChartData: [
          {
            legend: 'Legend 1',
            data: [
              {
                x: 1,
                y: 10,
                callOutAccessibilityData: {
                  ariaLabel: 'Custom aria label',
                },
              },
            ],
          },
        ],
      },
    },
    container => {
      const points = getById(container, /circle/i);
      expect(points).toHaveLength(1);
      expect(points[0].getAttribute('aria-label')).toEqual('Custom aria label');
    },
  );
});
