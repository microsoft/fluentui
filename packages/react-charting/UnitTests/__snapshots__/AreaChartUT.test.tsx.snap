// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_getAriaLabel Should return the correct aria label for a point with formatted x value and xAxisCalloutData in IST timezone 1`] = `"Jan 1, 2022. Legend 1, 10."`;

exports[`_getAriaLabel Should return the correct aria label for a point with formatted x value and xAxisCalloutData in Pacific timezone 1`] = `"Jan 1, 2022. Legend 1, 10."`;

exports[`_getAriaLabel Should return the correct aria label for a point with formatted x value and xAxisCalloutData in UTC timezone 1`] = `"Jan 1, 2022. Legend 1, 10."`;

exports[`_getAriaLabel Should return the correct aria label for a point with formatted x value and y value in IST timezone 1`] = `"1/1/2022, 5:30:00 AM. Legend 1, 10."`;

exports[`_getAriaLabel Should return the correct aria label for a point with formatted x value and y value in Pacific timezone 1`] = `"12/31/2021, 4:00:00 PM. Legend 1, 10."`;

exports[`_getAriaLabel Should return the correct aria label for a point with formatted x value and y value in UTC timezone 1`] = `"1/1/2022, 12:00:00 AM. Legend 1, 10."`;

exports[`_getAriaLabel Should return the correct aria label for a point with formatted x value and yAxisCalloutData in IST timezone 1`] = `"1/1/2022, 5:30:00 AM. Legend 1, 10 units."`;

exports[`_getAriaLabel Should return the correct aria label for a point with formatted x value and yAxisCalloutData in Pacific timezone 1`] = `"12/31/2021, 4:00:00 PM. Legend 1, 10 units."`;

exports[`_getAriaLabel Should return the correct aria label for a point with formatted x value and yAxisCalloutData in UTC timezone 1`] = `"1/1/2022, 12:00:00 AM. Legend 1, 10 units."`;
